src/main/kotlin/org/jitsi/nlj/transform/node/incoming/PaddingTermination.kt

```diff
@@ -15,35 +15,41 @@
  */
 package org.jitsi.nlj.transform.node.incoming
 
-import java.util.Collections
-import java.util.TreeMap
 import org.jitsi.nlj.PacketInfo
 import org.jitsi.nlj.stats.NodeStatsBlock
-import org.jitsi.nlj.transform.node.FilterNode
+import org.jitsi.nlj.transform.node.TransformerNode
 import org.jitsi.rtp.rtp.RtpPacket
-import org.jitsi.utils.LRUCache
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.utils.logging2.createChildLogger
+import kotlin.math.max
 
-class PaddingTermination : FilterNode("Padding termination") {
-    private val replayContexts: MutableMap<Long, MutableSet<Int>> = TreeMap()
-    private var numPaddingPacketsSeen = 0
+class PaddingTermination(parentLogger: Logger) : TransformerNode("Padding termination") {
+    private val logger = createChildLogger(parentLogger)
+    private var numPaddedPacketsSeen = 0
+    private var numPaddingOnlyPacketsSeen = 0
 
-    override fun accept(packetInfo: PacketInfo): Boolean {
+    override fun transform(packetInfo: PacketInfo): PacketInfo? {
         val rtpPacket = packetInfo.packetAs<RtpPacket>()
-        val replayContext = replayContexts.computeIfAbsent(rtpPacket.ssrc) {
-            Collections.newSetFromMap(LRUCache(1500))
-        }
 
-        return if (replayContext.add(rtpPacket.sequenceNumber)) {
-            true
-        } else {
-            numPaddingPacketsSeen++
-            false
+        if (rtpPacket.hasPadding) {
+            val paddingSize = rtpPacket.paddingSize
+            rtpPacket.length = max(rtpPacket.length - paddingSize, rtpPacket.headerLength)
+            rtpPacket.hasPadding = false
+            packetInfo.resetPayloadVerification()
+            numPaddedPacketsSeen++
+            if (rtpPacket.payloadLength == 0) {
+                numPaddingOnlyPacketsSeen++
+                packetInfo.shouldDiscard = true
```

Comment:
So is the idea to separate the 'deciding if a packet should be dropped' logic from the 'stream rewrite' logic (now in the `DiscardableDiscarder`)?

Reply:
Yeah, exactly.  In particular, so that padding removal happens before the a/v split, but dropping happens after retransmission requester.